flybywiresim/aircraft
# fix: Ground speed mini limits fix

This screenshot from Discord shows that the current implementation of ground speed mini originally added by https://github.com/flybywiresim/a32nx/pull/2304 has a bug such that Vapp is not being limited to a maximum of the VFE of the next higher flap configuration. It is using the VFE of the next **lower** flap configuration.
![image](https://user-images.githubusercontent.com/70166617/154778942-23332122-cc8a-4b71-841c-ef52fd0bf261.png)

## Summary of Changes

This PR corrects the bug identified above so that the Vapp targeted by ground speed mini is limited to a maximum of VFE-5 for CONF FULL.
## Screenshots (if necessary)


Under this PR, the maximum target airspeed will correctly be limited to VFE-5 in CONF FULL, and VFE of the next higher flap configuration (i.e., the next flap configuration that would be chosen in the approach) for CONF 1, 2, 3, and UP.


## References

The ground speed mini PR referenced above has the reference shown for the maximum ground speed mini target Vapp of VFE-5 limitation for CONF FULL and VFE of the next flap configuration for other than CONF FULL.



## Additional context

The GS mini function adds speed to the target VAPP when the current headwind component is higher than the tower headwind component that has been entered in the PERF APPR page of the MCDU. It adds 1 knot to VAPP for every 3 knots the current headwind is above the tower headwind. The higher target speed is capped at VFE-5 in CONF FULL, and VFE of the next higher flap configuration in CONF UP, 1, and 2.

The VFE speeds are here:
![Screenshot (1603)_LI](https://user-images.githubusercontent.com/70166617/154833671-9fcc792b-438f-4543-8a2b-b85b3a7d6931.jpg)
(Note that VFE next for Flaps UP will be 230, the VFE speed for CONF 1.)

Discord username (if different from GitHub):

## Testing instructions

I am a little unsure how much specific testing is needed as I have already done the following testing. Perhaps QA should be concentrated on making sure that nothing was broken by this PR. But here is how to test that the function works properly:
1. Activate the approach phase in the MCDU.
2. Fly managed speed with autothrust on (or track the managed speed LVAR in the aircraft behaviors - local variables in dev mode).
3. Set the winds to 0 in the PERF APPR page. (Setting a tailwind does not help here as only the headwind component matters).
4. In each flap configuration, verify that the target airspeed denoted by the magenta triangle next to the PFD speed tape does not go to a speed higher than VFE-5 for CONF FULL, or VFE of the next higher flap configuration for Flaps UP, CONF 1,2, and 3. That is, the maximum target airspeed for CONF FULL is 172 KCAS, CONF 3 is 177 KCAS, CONF 2 is 185 KCAS, CONF 1 is 200 KCAS, and Flaps UP is 230 KCAS. (Optionally, you can also verify that without this PR, the target speeds can increase all the way to the next lower flap configuration.)
5. Using the MSFS weather UI progressively add headwind until the target airspeed reaches the maximum value above and verify that further increases in headwind do not increase the target speed. Use the ND display of wind speed for the actual wind speed as the wind can vary quite a bit from what is input in the MSFS weather UI as that UI is bugged. It is suggested to set the gusts to zero to avoid large variations in wind speed. Increase the wind speed very gradually, allowing the airplane to stabilize as you increase the wind or you won't like what happens!
6. You can pre-calculate the headwind that would be needed to reach the maximum target airspeed for each configuration. It is (the target airspeed minus the approach speed) times 3. So, it is pretty difficult to reach the maximum target airspeed for Flaps UP. Even with 174 pax and a max fuel load, VAPP is only 148. So the headwind value needed to reach 230 KCAS is 246 knots. It can be done, but that's quite a wind!  And you need to go above that to verify that the target airspeed doesn't increase above 230 KCAS. (The CONF 3 VAPP is 160, so that will reduce the needed wind by 36 knots :-).)
7. Do a flight with a full approach to verify that all managed speed targets still work properly.

Be aware that the winds input in the MSFS weather UI is bugged. You can verify the actual winds either by looking at the wind shown in the A32NX ND (or by using Simvar watcher).

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page
